src/librustc/ty/maps/plumbing.rs

```diff
@@ -344,6 +344,48 @@ macro_rules! define_maps {
                 }
             }
 
+            /// Ensure that either this query has all green inputs or been executed.
+            /// Executing query::ensure(D) is considered a read of the dep-node D.
+            ///
+            /// This function is particularly useful when executing passes for their
+            /// side-effects -- e.g., in order to report errors for erroneous programs.
+            ///
+            /// Note: The optimization is only available during incr. comp.
+            pub fn ensure(tcx: TyCtxt<'a, $tcx, 'lcx>, key: $K) -> () {
+                let dep_node = Self::to_dep_node(tcx, &key);
+
+                // Ensuring an "input" or anonymous query makes no sense
+                assert!(!dep_node.kind.is_anon());
+                assert!(!dep_node.kind.is_input());
+                use dep_graph::DepNodeColor;
+                match tcx.dep_graph.node_color(&dep_node) {
+                    Some(DepNodeColor::Green(dep_node_index)) => {
+                        profq_msg!(tcx, ProfileQueriesMsg::CacheHit);
+                        tcx.dep_graph.read_index(dep_node_index);
+                    }
+                    Some(DepNodeColor::Red) => {
+                        let _ = tcx.$name(key);
+                    }
+                    None => {
+                        // Huh
+                        if !tcx.dep_graph.is_fully_enabled() {
+                            let _ = tcx.$name(key);
+                            return;
+                        }
```

Comment:
@nikomatsakis Open question: Should the preceding 4 lines be in `try_mark_green`?

Reply:
For reference, `try_mark_green` will `unwrap` a `None` [here](https://github.com/rust-lang/rust/blob/a47c9f870f13603a06ffe63ab4834fc716912843/src/librustc/dep_graph/graph.rs#L435) if `tcx.dep_graph.is_fully_enabled()` would return False.

Reply:
I'd rather keep the explicit `panic` in `try_mark_green`, I think.

Reply:
I'll leave the code here as it is then.

Reply:
I also expected those lines to be in `try_mark_green`, but I trust @michaelwoerister's intuitions here. =) That said, I think we could consider trying to extract some of the common code between this function and `try_get_with`. I had thought it might make sense to extract a helper function that returns `Some(dep_node_index)` if either: the node color is green or can be made green, and `None` otherwise. This function here could invoke `tcx.$name(key)` if `None` is returned, and `try_get_with` can do whatever it does now.

Reply:
I like it! I'll implement that refactor as a follow on PR, unless this round of testing does not go well and I have to fix something.